### PR TITLE
{numlib}[foss/2020a,intel/2020a] preCICE v2.2.0 w/ Python 3.8.2

### PR DESCRIPTION
**Pull request by besserox**
(created using `eb --new-pr`)

### CODE REVIEW

**Comment by besserox** [comment]
One additional comment:
The tests are disabled with the intel toolchain because they fail.
This is being discussed with the preCICE developers https://github.com/precice/precice/issues/966

File: easybuild/easyconfigs/p/preCICE/preCICE-2.2.0-intel-2020a-Python-3.8.2.eb [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+# Make sure MPI support is enabled
+configopts = '-DPRECICE_MPICommunication=ON '
+# Make sure PETSc support is enabled
+configopts += '-DPRECICE_PETScMapping=ON '
+# Make sure PythonActions support is enabled
+configopts += '-DPRECICE_PythonActions=ON '
+# force Eigen include directory which is not found correctly otherwise
+configopts += '-DEIGEN3_INCLUDE_DIR="$EBROOTEIGEN/include"'
+
+# Run basic tests -> disabled because they fail with intel toolchain
```
Comment (boegel) [inline easybuild/easyconfigs/p/preCICE/preCICE-2.2.0-intel-2020a-Python-3.8.2.eb:47]: @besserox Please add a pointer to https://github.com/precice/precice/issues/966 as a comment here too?

**Review by boegel — CHANGES_REQUESTED** [review]

**Comment by boegel** [comment]
@boegelbot please test @ generoso

File: easybuild/easyconfigs/p/preCICE/preCICE-2.2.0-intel-2020a-Python-3.8.2.eb (hunk 1 above)
Comment (besserox) [reply easybuild/easyconfigs/p/preCICE/preCICE-2.2.0-intel-2020a-Python-3.8.2.eb]: Done

**Comment by besserox** [comment]
> Test report by @boegelbot
> **FAILED**
> Build succeeded for 0 out of 2 (2 easyconfigs in total)
> generoso-c1-s-1 - Linux centos linux 8.2.2004, x86_64, Intel(R) Xeon(R) CPU E5-2667 v3 @ 3.20GHz (haswell), Python 3.6.8
> See https://gist.github.com/f2a6f5d63d334b51d6f4bf7f3011cee8 for a full test report.

@boegel I'm a bit confused by this error here
```
CMAKE_PREFIX_PATH
   /shared/easybuilder/CentOS8/haswell/software/PETSc/3.12.4-foss-2020a-Python-3.8.2
-- Checking for module 'PETSc>=3.12'
--   Found PETSc, version 3.12.4
CMake Error at /shared/easybuilder/CentOS8/haswell/software/CMake/3.16.4-GCCcore-9.3.0/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:146 (message):
  Could NOT find PETSc (missing: PETSc_INCLUDE_DIRS) (found suitable version
  "3.12.4", minimum required is "3.12")
Call Stack (most recent call first):
  /shared/easybuilder/CentOS8/haswell/software/CMake/3.16.4-GCCcore-9.3.0/share/cmake-3.16/Modules/FindPackageHandleStandardArgs.cmake:393 (_FPHSA_FAILURE_MESSAGE)
  cmake/modules/FindPETSc.cmake:229 (find_package_handle_standard_args)
  CMakeLists.txt:216 (find_package)
```

I tested on 2 different clusters and CMake has no problem finding PETSc
```
CMAKE_PREFIX_PATH
   /home/xbesseron/.local/easybuild/software/PETSc/3.12.4-intel-2020a-Python-3.8.2
-- Checking for module 'PETSc>=3.12'
--   Found PETSc, version 3.12.4
-- Found PETSc: TRUE (found suitable version "3.12.4", minimum required is "3.12") 
```

Any idea where to look at?
I could try force the value of `PETSc_INCLUDE_DIRS` but it seems it will just hide the real issue.

**Comment by lexming** [comment]
Test report by @lexming
**SUCCESS**
Build succeeded for 10 out of 10 (2 easyconfigs in total)
node371.hydra.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6148 CPU @ 2.40GHz, Python 2.7.5
See https://gist.github.com/a05b113962fcde541fa45eac78aeca15 for a full test report.

**Review by lexming — COMMENTED** [review]
@besserox The failed test from @boegel might be caused by the `pkg-config` from the host system in CentOS 8, which ships a quite particular version. I've seen such errors before. Could you add `pkg-config` v0.29.2 from EB as build dependency?

If this does not help, explicitly setting the paths to libraries and headers in CMake is not a bad solution. This guarantees that any PETSc in the host system (if present) will not be favoured over the PETSc from EB.

**Comment by boegel** [comment]
Adding `pkg-config` as a build dependency doesn't help, I just tested that.

I'm trying to figure out why CMake first claims to have found the `PETSc` installation, to then go and bitch about not finding the header files (which are there in the installation it finds). I can't seem to find any relevant output in the CMake output files (but that doesn't really surprise me).

**Comment by besserox** [comment]
Just tried on my side to add `pkg-config` as a build dependency. 
With that, I can reproduce the error on the cluster I'm using (CentOS 7.8).

So I can debug it.

**Comment by boegel** [comment]
@besserox Hold on, *adding* pkg-config triggers the issue for you?! Oh boy...

There's definitely some unholy combo of `pkg-config` and `CMake` going on in `preCICE`, see https://github.com/precice/precice/commit/1c3dad78cde8f6eb5198a522cd6d067bc3cf2602, and they've had trouble with it before, see https://github.com/precice/precice/issues/888

**Comment by besserox** [comment]
Some progress with this bug:
- The working setup uses the system `pkg-config` version 0.27.1
- The failing setup uses the EasyBuild `pkg-config` version 0.29.2

It's the only difference.

What happens here is that:
- For the working version
`pkg-config --cflags-only-I PETSc` returns `-I/work/.../software/PETSc/3.12.4-foss-2020a-Python-3.8.2/include` 
- For the failing version
`pkg-config --cflags-only-I PETSc` returns `""` (empty)
Because of that, CMake is not able to find the `PETSc_INCLUDE_DIRS`, and it fails.

The reason of this behavior in `pkg-config` 0.29.2 is a bug fix that removes duplicates from the `CPATH` environment variable in order to match the behavior of GCC (https://github.com/frida/pkg-config/commit/4ade8fc90115c4932bc198a500a1efd13f243cd4). But this is probably not what CMake wants.

As a summary, I would say it is an incompatibility with `CMake 3.16.4` and `pkg-config 0.29.2`.
I don't know if it is fixed in more recent version of CMake.


As a side note, the result of `pkg-config` is also 'cleaned up' is the same way for `libxml2`. But that does not seem to impact CMake, but we should probably verify if the right libraries and include paths are picked up.

**Comment by boegel** [comment]
@boegelbot please test @ generoso

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node3524.doduo.os - Linux RHEL 8.2, x86_64, AMD EPYC 7552 48-Core Processor (zen2), Python 3.6.8
See https://gist.github.com/c87b72a64163b8202d2ab58845971033 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node3155.skitty.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) Gold 6140 CPU @ 2.30GHz, Python 3.6.8
See https://gist.github.com/92246fce86c29b04b308c17c5d5ec817 for a full test report.

**Comment by boegel** [comment]
Test report by @boegel
**SUCCESS**
Build succeeded for 2 out of 2 (2 easyconfigs in total)
node2703.swalot.os - Linux centos linux 7.9.2009, x86_64, Intel(R) Xeon(R) CPU E5-2660 v3 @ 2.60GHz (haswell), Python 3.6.8
See https://gist.github.com/7bcc48dc35e05ed868d2569aef93fe2d for a full test report.

**Review by boegel — APPROVED** [review]

**Comment by boegel** [comment]
@besserox Oh my, what a mess... Thanks a lot of figuring this out!

**Comment by boegel** [comment]
Going in, thanks @besserox!

**Pull request merged**